rule_builder.py

```diff
@@ -0,0 +1,2069 @@
+        return Or(self, other)
+
+    def __lshift__(self, other: Iterable[OptionFilter[Any]]) -> "Rule[TWorld]":
+        """Convenience operator to filter an existing rule with an option filter"""
+        return Filtered(self, options=other)
+
+    def __bool__(self) -> Never:
+        """Safeguard to prevent devs from mistakenly doing `rule1 and rule2` and getting the wrong result"""
+        raise TypeError("Use & or | to combine rules, or use `is not None` for boolean tests")
+
+    @override
+    def __str__(self) -> str:
+        options = f"options={self.options}" if self.options else ""
+        return f"{self.__class__.__name__}({options})"
+
+    @classmethod
+    def __init_subclass__(cls, /, game: str) -> None:
+        if game != "Archipelago":
+            custom_rules = CustomRuleRegister.custom_rules.setdefault(game, {})
+            if cls.__qualname__ in custom_rules:
+                raise TypeError(f"Rule {cls.__qualname__} has already been registered for game {game}")
+            custom_rules[cls.__qualname__] = cls
+        elif cls.__module__ != "rule_builder":
+            # TODO: test to make sure this works on frozen
+            raise TypeError("You cannot define custom rules for the base Archipelago world")
+        cls.game_name = game
+
+    class Resolved(metaclass=CustomRuleRegister):
+        """A resolved rule for a given world that can be used as an access rule"""
+
+        _: dataclasses.KW_ONLY
+
+        player: int
+        """The player this rule is for"""
+
+        caching_enabled: bool = dataclasses.field(repr=False, default=True, kw_only=True)
+        """If the world this rule is for has caching enabled"""
+
+        force_recalculate: ClassVar[bool] = False
+        """Forces this rule to be recalculated every time it is evaluated.
+        Forces any parent composite rules containing this rule to also be recalculated. Implies skip_cache."""
+
+        skip_cache: ClassVar[bool] = False
+        """Skips the caching layer when evaluating this rule.
+        Composite rules will still respect the caching layer so dependencies functions should be implemented as normal.
+        Set to True when rule calculation is trivial."""
+
+        always_true: ClassVar[bool] = False
+        """Whether this rule always evaluates to True, used to short-circuit logic"""
+
+        always_false: ClassVar[bool] = False
+        """Whether this rule always evaluates to True, used to short-circuit logic"""
+
+        def __post_init__(self) -> None:
+            object.__setattr__(
+                self,
+                "caching_enabled",
+                self.caching_enabled and not self.force_recalculate and not self.skip_cache,
+            )
+
+        def __call__(self, state: CollectionState) -> bool:
+            """Evaluate this rule's result with the given state, using the cached value if possible"""
+            if not self.caching_enabled:
+                return self._evaluate(state)
```

Comment:
I'm being a bit anal about performance here, but what do you think of the idea of having a `CachedRule` subclass, then just doing `__call__ = _evaluate` in this class, and then overriding `__call__` in `CachedRule` with what you've got here?